include/range/v3/utility/iterator.hpp

```diff
@@ -619,6 +619,64 @@ namespace ranges
             Char const *delim_;
         };
 
+        template <typename Delim, typename Char = char,
+                  typename Traits = std::char_traits<Char>>
+        struct ostream_joiner
+        {
+            using difference_type = std::ptrdiff_t;
+            using char_type       = Char;
+            using traits_type     = Traits;
+            using ostream_type    = std::basic_ostream<Char, Traits>;
+
+            constexpr ostream_joiner() = default;
+            ostream_joiner(ostream_type &s, Delim const &d)
+              : delim_(d), sout_(std::addressof(s)), first_(true)
+            {}
+            ostream_joiner(ostream_type &s, Delim &&d)
+              : delim_(std::move(d)), sout_(std::addressof(s)), first_(true)
+            {}
+            template <typename T>
+            ostream_joiner& operator=(T const &value)
+            {
+                RANGES_EXPECT(sout_);
+                if (!first_)
+                    *sout_ << delim_;
+                first_ = false;
+                *sout_ << value;
+                return *this;
+            }
+            ostream_joiner& operator*() noexcept
+            {
+                return *this;
+            }
+            ostream_joiner& operator++() noexcept
+            {
+                return *this;
+            }
+            ostream_joiner& operator++(int) noexcept
+            {
+                return *this;
+            }
+          private:
+            Delim delim_;
+            ostream_type *sout_;
+            bool first_;
+        };
+
+        struct make_ostream_joiner_fn
+        {
+            template <typename Delim, typename Char, typename Traits>
```

Comment:
`Delim` needs to model `SemiRegular` in order for the iterator to do so.; we should constrain this:
```c++
template <typename Delim, typename Char, typename Traits,
    CONCEPT_REQUIRES_(SemiRegular<Delim>())>
```